File: arrays/3Sum/3Sum.ts
export class Solution {
    /**
    * @param {string} testInput
    * @return {any} Returns either boolean to validate the test cases or a custom result. 
    */
    run_solution(testInput: any) {
        const sorted: number[] = testInput.sort((a, b) => a - b);
        const result: number[][] = []
        
        for (let i =0; i<sorted.length; i++){
            let left = i+1;
            let right = sorted.length -1;

            while(left < right){
                const sum = sorted[i] + sorted[left] + sorted[right];

                if(sum>0) {
                    right--;
                } else if(sum<0){
                    left++;
                } else {
                    result.push([sorted[i], sorted[left], sorted[right]])
                    left++;
                    right++
                    while(left < right && sorted[left] === sorted[left -1]){
                        left++
                    }
                }
            }
        }
        return result
        console.log(sorted)
    }
}

File: arrays/3Sum/3Sum.py
"""Solution Module"""


class Solution:
    """Solution Class"""

    def run_solution(self, test_input):
        """Solution Method"""
        sorted_arr = sorted(test_input)
        result: [[]] = []

        for idx, val in enumerate(sorted_arr):
            if val > 0:
                break

            left = idx + 1
            right = len(sorted_arr) - 1

            while left < right:
                sumed = val + sorted_arr[left] + sorted_arr[right]

                if sumed > 0:
                    right -= 1
                elif sumed < 0:
                    left += 1
                else:
                    result.append([val, sorted_arr[left], sorted_arr[right]])
                    right -= 1
                    left += 1
                    while left < right:
                        left += 1
        return result
